server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryServer.java

```diff
@@ -228,10 +228,15 @@ public void onNext(DiscoveryRequest request) {
 
       if (defaultTypeUrl.equals(ANY_TYPE_URL)) {
         if (requestTypeUrl.isEmpty()) {
-          responseObserver.onError(
-              Status.UNKNOWN
-                  .withDescription(String.format("[%d] type URL is required for ADS", streamId))
-                  .asRuntimeException());
+          synchronized (responseObserver) {
+            if (!isClosing) {
+              isClosing = true;
+              responseObserver.onError(
+                  Status.UNKNOWN
+                      .withDescription(String.format("[%d] type URL is required for ADS", streamId))
+                      .asRuntimeException());
+            }
+          }
```

Comment:
Can we use
```
closeWithError(Status.UNKNOWN .withDescription...)
```
I see that `closeWithError` also cancels the stream, but this is an exceptional situation anyway.

Reply:
sure